README.md

```diff
@@ -956,6 +961,117 @@ the opposite order:
 
 ## Miscellaneous
 
+### Traversing within [`@optional`](#optional)
+Including an optional statement in GraphQL has no performance issues on its own,
+but if you continue traversing within an optional there may be significant performance implications.
+
+Going forward, we will refer to two different kinds of `@optional` directives.
+
+- A *"simple"* optional is a vertex with an `@optional` directive that does not expand
+any vertex fields within it.
+For example:
+```
+{
+    Animal {
+        name @output(out_name: "name")
+        in_Animal_ParentOf @optional {
+            name @output(out_name: "parent_name")
+        }
+    }
+}
+```
+OrientDB `MATCH` currently allows the last step in any traversal to be optional.
+Therefore, the equivalent `MATCH` traversal for the above `GraphQL` is as follows:
+```
+MATCH {
+    class: Animal,
+    as: Animal___1
+}.in('Animal_ParentOf') {
+    as: Animal__in_Animal_ParentOf___1
+}
+```
+
+- A *"compound"* optional is a vertex with an `@optional` directive which does expand
+vertex fields within it.
+For example:
+```
+{
+    Animal {
+        name @output(out_name: "name")
+        in_Animal_ParentOf @optional {
+            name @output(out_name: "parent_name")
+            in_Animal_ParentOf {
+                name @output(out_name: "grandparent_name")
+            }
+        }
+    }
+}
+```
+Currently, this cannot represented by a simple `MATCH` query.
+Specifically, the following is *NOT* a valid `MATCH` statement,
+because the optional traversal follows another edge:
+```
+MATCH {
+    class: Animal,
+    as: Animal___1
+}.in('Animal_ParentOf') {
+    optional: true,
+    as: Animal__in_Animal_ParentOf___1
+}.in('Animal_ParentOf') {
+    as: Animal__in_Animal_ParentOf__in_Animal_ParentOf___1
+}
+```
+
+Instead, we represent a *compound* optional by taking an union (`UNIONALL`) of two distinct
+`MATCH` queries. For instance, the `GraphQL` query above can be represented as follows:
+```
+LET
+$match1 = (
+    MATCH {
+        class: Animal,
+        as: Animal___1,
+        where: (
+            (in_Animal_ParentOf IS null)
+            OR
+            (in_Animal_ParentOf.size() = 0)
+        ),
+    }
+),
+$match2 = (
+    MATCH {
+        class: Animal,
+        as: Animal___1
+    }.in('Animal_ParentOf') {
+        as: Animal__in_Animal_ParentOf___1
+    }.in('Animal_ParentOf') {
+        as: Animal__in_Animal_ParentOf__in_Animal_ParentOf___1
+    }
+),
+$final_match = UNIONALL($match1, $match2)
+```
+In the first case where the optional edge is not followed,
+we have to explicitly filter out all vertices where the edge *could have been folowed*.
+This is to eliminate duplicates between the two `MATCH` selections.
+
+The previous example is not *exactly* how we implement *compound* optionals
+(we also have `SELECT` statements within `$match1` and `$match2`), but it illustrates the point.
+
+#### Performance Penalty
+
+If we have many *compound* optionals in the given `GraphQL`,
```

Comment:
Could you re-examine how you use bullets here, please? I think you can leverage having parallel structure for sentences here. e.g.

```
- For each of the 2^n subsets....
- For each traverse `t`
```

Not sure how much it would affect meaning, though... Try it out?